### PR TITLE
Hyperparam optimization

### PR DESCRIPTION
**Pull request by delhomer**
This PR introduces a new module (`param_optim.py`) that allows to launch several runs of training for a same tuple model (feature detection, semantic segmentation) - dataset (shapes, mapillary, ...), with different hyperparameters (batch size, learning rate, learning rate decay, dropout rate, or even network architecture).
During the hyperparameter optimization, each instance configuration produces a model backup on the file system (`datapath/dataset/output/model/checkpoints/instance/backup.h5`), as this would be done on a standard training process. In addition to that, at the end of the optimization process, a "best model" is stored one the file system as well (`datapath/dataset/output/model/checkpoints/best-model-imsize-aggregationpolicy.h5`), this backup coming with an instance description in `.json` format.

The hyperparameter optimization process does not take into account image size and label aggregation policy (`aggregate` *vs* `full`). That's why these parameters appear to characterize the best model files (and as a consequence, several "best models" may exist).

In further developments (in particular when using training results), the best models will be used by default to infer on testing image labels.

Fix issue #28 : the code provides the functions to do hyperparameter optimization, this process will be run as a background task in the next weeks.

### CODE REVIEW

File: deeposlandia/paramoptim.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                        default=18000,
+                        help=("Number of training images"))
+    parser.add_argument('-iv', '--nb-validation-image',
+                        type=int,
+                        default=2000,
+                        help=("Number of validation images"))
+    return parser
+
+def get_data(folders, dataset, model, image_size, batch_size):
+    """
```
Comment (garaud) [inline deeposlandia/paramoptim.py]: Missing docstring.

File: deeposlandia/paramoptim.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                           "Please generate a valid dataset before calling the training program."))
+        sys.exit(1)
+    nb_labels = len(label_ids)
+    return nb_labels, train_generator, validation_generator, test_generator
+
+def run_model(train_generator, validation_generator, output_folder,
+              instance_name, image_size, aggregate_value, nb_labels,
+              nb_training_image, nb_validation_image,
+              batch_size, dropout, network, learning_rate, learning_rate_decay):
+    """
```
Comment (garaud) [inline deeposlandia/paramoptim.py]: missing docstring.

File: deeposlandia/paramoptim.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    else:
+        utils.logger.error(("Unrecognized model. Please enter 'feature_detection' "
+                            "or 'semantic_segmentation'."))
+        sys.exit(1)
+    model = Model(net.X, net.Y)
+    opt = Adam(lr=learning_rate, decay=learning_rate_decay)
+    model.compile(loss=loss_function, optimizer=opt, metrics=['acc'])
+
+    # Model training
+    STEPS = nb_training_image // batch_size
```
Comment (garaud) [inline deeposlandia/paramoptim.py]: I think this is not necessary to have variable names in uppercase.

File: deeposlandia/paramoptim.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                               validation_data=validation_generator,
+                               validation_steps=VAL_STEPS,
+                               callbacks=[checkpoints, earlystop, terminate_on_nan])
+    ref_metric = max(hist.history['val_acc'])
+    return {'model': model, 'val_acc': ref_metric,
+            'batch_size': batch_size, 'network': network, 'dropout': dropout,
+            'learning_rate': learning_rate, 'learning_rate_decay': learning_rate_decay}
+
+if __name__=='__main__':
+    """Main method: run a convolutional neural network using Keras API
```
Comment (garaud) [inline deeposlandia/paramoptim.py]: Maybe turn the string into a comment. This is not a function docstring.

Comment (garaud) [reply deeposlandia/paramoptim.py]: Or you can move this content into a really `main` function.

**Review by garaud — CHANGES_REQUESTED** [review]
I push a tiny commit and have some comments.

File: deeposlandia/paramoptim.py (hunk 4 above)
Comment (delhomer) [reply deeposlandia/paramoptim.py]: I've finally dropped it: the information already appears in the module docstring.

File: deeposlandia/paramoptim.py (hunk 3 above)
Comment (delhomer) [reply deeposlandia/paramoptim.py]: Both affected variables have been lower-cased.

File: deeposlandia/paramoptim.py (hunk 2 above)
Comment (delhomer) [reply deeposlandia/paramoptim.py]: Added.

File: deeposlandia/paramoptim.py (hunk 1 above)
Comment (delhomer) [reply deeposlandia/paramoptim.py]: Added.

**Review by garaud — APPROVED** [review]

**Comment by garaud** [comment]
looks good for me.
